File: go/cmd/pkg/shell_sort.go
package pkg

import (
	"fmt"
	"time"
)

// timeDecorator оборачивает функцию и замеряет её время выполнения
func timeDecorator(f func([]int) []int) func([]int) []int {
	return func(arr []int) []int {
		start := time.Now()           // Засекаем время начала
		result := f(arr)              // Выполняем функцию
		duration := time.Since(start) // Засекаем время окончания
		fmt.Printf("Время выполнения функции: %v мс\n", duration.Milliseconds())
		return result
	}
}

// shellSort сортировка Шелла
func shellSort(arr []int) []int {
	n := len(arr)
	gap := 1

	// Вычисляем начальный промежуток по формуле Кнута
	for gap < n/3 {
		gap = 3*gap + 1
	}

	// Последовательно уменьшаем промежуток
	for gap > 0 {
		// Применяем сортировку вставками для текущего промежутка
		for i := gap; i < n; i++ {
			temp := arr[i]
			j := i
			// Сдвигаем элементы, пока не найдем позицию для вставки
			for j >= gap && arr[j-gap] > temp {
				arr[j] = arr[j-gap]
				j -= gap
			}
			arr[j] = temp
		}
		// Уменьшаем промежуток
		gap = gap / 3
	}

	return arr
}

File: py/sorting/shell_sort.py
import time

def time_decorator(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()  # Засекаем время до выполнения
        result = func(*args, **kwargs)  # Выполняем функцию
        end_time = time.time()  # Засекаем время после выполнения
        print(f"Время выполнения функции {func.__name__}: {(end_time - start_time) * 1000} мс")
        return result
    return wrapper


@time_decorator
def shell_sort(arr):
    n = len(arr)
    gap = 1
    
    while gap < n // 3:
        gap = 3 * gap + 1
    
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]
            j = i
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap
            arr[j] = temp
        gap = gap // 3
    
    return arr